lib/Builtins.cpp

```diff
@@ -14,171 +14,335 @@
+////  Mangled name parsing utilities
+////////////////////////////////////////////////////////////////////////////////
+
+// capture real name of function and struct types
+std::string GetUnmangledName(const std::string &str, size_t *pos) {
+  char *end;
+  int name_len = strtol(&str[*pos], &end, 10);
+  if (!name_len) {
+    return "";
+  }
+  size_t name_pos = end - str.data();
+
+  *pos = name_pos + name_len;
+  return str.substr(name_pos, name_len);
+}
+
+// capture parameter type and qualifiers based on Itanium ABI with OpenCL types
+// - return new parsing position pos, or zero for error
+size_t GetParameterType(const std::string &mangled_name,
+                        clspv::Builtins::ParamTypeInfo *type_info, size_t pos) {
+  // Parse a parameter type encoding
+  char type_code = mangled_name[pos++];
+
+  int blen = 1;
+  switch (type_code) {
+    // qualifiers
+  case 'P': // Pointer
+  case 'R': // Reference
+    return GetParameterType(mangled_name, type_info, pos);
+  case 'k': // ??? not part of cxxabi
+  case 'K': // const
+  case 'V': // volatile
+    return GetParameterType(mangled_name, type_info, pos);
+  case 'U': { // Address space
+    std::string address_space = GetUnmangledName(mangled_name, &pos);
+    return GetParameterType(mangled_name, type_info, pos);
+  }
+    // OCL types
+  case 'D':
+    type_code = mangled_name[pos++];
+    if (type_code == 'v') { // OCL vector
+      char *end;
+      int numElems = strtol(&mangled_name[pos], &end, 10);
+      if (!numElems) {
+        return 0;
+      }
+      type_info->vector_size = numElems;
+      pos = end - mangled_name.data();
+
+      if (mangled_name[pos++] != '_') {
+        return 0;
+      }
+      return GetParameterType(mangled_name, type_info, pos);
+    } else if (type_code == 'h') { // OCL half
+      type_info->type_id = Type::FloatTyID;
+      type_info->is_signed = true;
+      type_info->byte_len = 2;
+    } else {
+#ifdef DEBUG
+      printf("Func: %s\n", mangled_name.c_str());
+      llvm_unreachable("failed to demangle name");
+#endif
+      return 0;
+    }
+    break;
+
+    // element types
+  case 'l':
+    blen <<= 1; // long
+  case 'i':
+    blen <<= 1; // int
+  case 's':
+    blen <<= 1; // short
+  case 'c':     // char
+  case 'a':     // signed char
+    type_info->type_id = Type::IntegerTyID;
+    type_info->is_signed = true;
+    type_info->byte_len = blen;
+    break;
+  case 'm':
+    blen <<= 1; // unsigned long
+  case 'j':
+    blen <<= 1; // unsigned int
+  case 't':
+    blen <<= 1; // unsigned short
+  case 'h':     // unsigned char
+    type_info->type_id = Type::IntegerTyID;
+    type_info->is_signed = false;
+    type_info->byte_len = blen;
+    break;
+  case 'd':
+    blen = 2; // double float
+  case 'f':
+    blen <<= 2; // single float
+    type_info->type_id = Type::FloatTyID;
+    type_info->is_signed = true;
+    type_info->byte_len = blen;
+    break;
+  case 'v': // void
+    break;
+  case '1':
```

Comment:
Please explain this.  And why not  '0'

For example, it's not very intuititve what's going on, even when staring at the mangling reference
https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling


I *think* what's going on is that this is a typedef'd name?  
E.g.   struct Foobar { ... }
would have  6Foobar in it, and these cases occur when you're looking at that initial digit. 

??  That's interesting, and needs explanation.

Is it always going to be a struct?  E.g. what if it's a typedef'd name for different kind of thing .like pointer-to-int-in-private-space. ?

Reply:
For unqualified names: https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangle.unqualified-name
It gets more complicated for scoped names, but we shouldn't see that for builtins. 

This doesn't check for '0' since it is only looking for the beginning of a \<positive length number>.

Reply:
typedefs should be resolved in mangled names.

Reply:
I added abit more clarification in the comments.